### PR TITLE
Use presence of _symint in kernel name to generate symint sig or not

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84579

Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key.  This changes it so that SymInt or not in type is based on whether
or not you have _symint in the name of the kernel or not.  This means
that even when we make operators support SymInt, you no longer have to
go and update all the preexisting definitions; instead, you now
selectively write _symint to opt individual kernels into SymInt support.

I then go and update a bunch of kernels that don't have proper SymInt
support to make use of this convention.  There is some hacking around
for view generation code.

I also add support for external backends to specify 'symint' operators, for which we generate SymInt signatures instead of regular signatures.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D39310060](https://our.internmc.facebook.com/intern/diff/D39310060)

### CODE REVIEW

**Comment by wconstab** [comment]
> Something people found confusing was that whether or not a native::
signature would get SymInt or not in its type was based on the dispatch
key

Before this PR, was the idea that 'meta' kernels were getting c++ symint signatures but real kernels weren't, and now this becomes more explicit because for either meta or composite kernels you just have to write _symint in the name and take out the guesswork?

**Comment by ezyang** [comment]
> Before this PR, was the idea that 'meta' kernels were getting c++ symint signatures but real kernels weren't, and now this becomes more explicit because for either meta or composite kernels you just have to write _symint in the name and take out the guesswork?

Yep

File: torchgen/api/types.py [hunk 1]
```diff
@@ -783,10 +783,12 @@ def kernel_signature(
         # not so good for more conventional backends but we don't have any of
         # those.  If we do, that's time to add a new Signature that is a cross
         # between DispatcherSignature and NativeSignature
-        assert backend_index.symint
         return DispatcherSignature.from_schema(f.func, prefix=prefix)
     else:
-        return NativeSignature(f.func, prefix=prefix, symint=backend_index.symint)
+        meta = backend_index.get_kernel(f)
+        return NativeSignature(
+            f.func, prefix=prefix, symint=meta is not None and meta.supports_symint()
```
Comment (wconstab) [inline torchgen/api/types.py:790]: does having a python decomp count here in favor of being symint compatible, or is a 'meta' kernel always strictly necessary?

File: torchgen/gen_functionalization_type.py [hunk 2]
```diff
@@ -84,21 +84,21 @@ def gen_composite_view_copy_kernel(g: NativeFunctionsViewGroup) -> Optional[str]
...
-  auto int_size = c10::asIntArrayRefSlow(size);
-  DimVector shape = infer_size_dv(int_size, self.numel());
+at::Tensor view_copy(const at::Tensor & self, at::IntArrayRef size) {
+  DimVector shape = infer_size_dv(size, self.numel());
   if (!at::detail::computeStride(self.sizes(), self.strides(), shape).has_value()) {
-    return self.reshape(int_size);
+    return self.reshape(size);
   } else {
-    auto output = at::_ops::view::call(self, size);
+    auto output = at::_ops::view::call(self, c10::SymIntArrayRef::fromIntArrayRef(size));
```
Comment (wconstab) [inline torchgen/gen_functionalization_type.py:92]: ok what's going on in this kernel? the `view_copy` name above didn't get `_symint` tacked on, and some of the symint related stuff got removed, but then we are constructing symints and invoking view.

File: torchgen/api/types.py (hunk 1 above)
Comment (ezyang) [reply torchgen/api/types.py:790]: Python decomp suffices, meta is not required

File: torchgen/native_function_generation.py [hunk 3]
```diff
@@ -303,7 +303,7 @@ def generate_function(
         func.name.unambiguous_name()
         if func.kind() == SchemaKind.out
         else cpp.name(func)
-    )
+    ) + "_symint"  # TODO: hmm
```
Comment (wconstab) [inline torchgen/native_function_generation.py:306]: why is _symint assumed here? shouldn't this be driven by the yaml?

Comment (ezyang) [reply torchgen/native_function_generation.py:306]: Sooo I am not entirely sure here, but the idea is that these kernels are autogenerated, and they handle SymInts, so we should give them symint names so they get the right signature. It's a bit dodgy though... I got here by just tweaking until it compiled

Comment (wconstab) [reply torchgen/native_function_generation.py:306]: that sounds reasonable

clarification: these 'generated' kernels are going to call into handwritten kernels (adapting from functional to out or vice versa) right?  if so, would we be potentially having a situation like this?
foo_out <-- handwritten, no symint support
foo_symint <-- generated, just trivially unwraps symints and expects ints, calls foo_out

if so, i still don't think that's a problem.  other than, we should probably make it clearer what's going on

**Review by wconstab — APPROVED** [review]
This sounds like a better policy to me, and it looks correct outside of the minor questions I raised

**Comment by ezyang** [comment]
I'm going to squish this with the earlier diff as it won't be possible to get XLA working without both sides of the coin.

File: torchgen/native_function_generation.py (hunk 3 above)
Comment (ezyang) [reply torchgen/native_function_generation.py:306]: Err, not really? The generated kernels typically call into other API as defined by the codegen; e.g., a generated out will generate a call to functional and then copy_ (but the functional call is not a direct native:: call; it'll just be to some dispatched api).

File: torch/csrc/jit/runtime/static/generated_ops.cpp [hunk 4]
```diff
@@ -1812,7 +1812,7 @@ REGISTER_OPERATOR_FUNCTOR(aten::prod, aten_prod, [](Node* n) -> SROperator {
       }
       auto& out = p_node->Output(0).toTensor();
       fastResizeToZero(out);
-      at::native::prod_out(self, dtype, out);
+      at::native::prod_out_symint(self, dtype, out);
```
Comment (ezyang) [inline torch/csrc/jit/runtime/static/generated_ops.cpp:1815]: This diff can be reverted if I only `_symint` autogen functions if the source schema has symint

File: torchgen/api/types.py [hunk 5]
```diff
@@ -778,15 +780,18 @@ def kernel_signature(
...
     # With external backends, we'd like to enforce that they write their kernels with schemas
     # that match the Dispatcher API directly, if they can.
+    meta = backend_index.get_kernel(f)
+    symint = meta is not None and meta.supports_symint()
+    # TODO: Would like to do this, but autogen is too sloppy right now
+    """
+    if symint:
+        assert f.func.has_symint(), \
+            f"attempted to define symint kernel for {backend_index.dispatch_key} without SymInt in schema"
+    """
```
Comment (ezyang) [inline torchgen/api/types.py:790]: Ditto here, less sloppy autogen will let this be restored

File: torchgen/gen.py [hunk 6]
```diff
@@ -606,7 +612,7 @@ def __call__(self, f: NativeFunction) -> str:
                     )
```
Comment (ezyang) [inline torchgen/gen.py:612]: static dispatch looks like it needs to be updated

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39310060).

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39310060).

**Comment by seemethere** [comment]
Rebased to include new xla hash from @zou3519, `f00dd2f35ecf6455d97237d63c70c9c8ec190940`

**Pull request closed without merging**